parlai/utils/torch.py

```diff
@@ -235,3 +237,305 @@ def forward(self, xs):
+        total number of parameters in the model.
+    """
+    return sum(p.numel() for p in model.parameters())
+
+
+def trainable_parameters(model: torch.nn.Module) -> int:
+    """
+    Count the total number of trainable parameters in the model.
+
+    :param model:
+        the model whose parameters we wish to count.
+
+    :return:
+        total number of trainable parameters in the model.
+    """
+    return sum(p.numel() for p in model.parameters() if p.requires_grad)
+
+
+Chunk = TypeVar('Chunk')
+
+
+PipelineWorkItem = namedtuple(
+    'PipelineWorkItem', ['chunk_idx', 'layer_nos', 'next_device']
+)
+
+
+class PipelineHelper(object):
+    """
+    PipelineHelper assists with implementing pipelining in model parallelism.
+
+    For a tutorial on model parallelism, as it's implemented in parts of ParlAI,
+    see https://pytorch.org/tutorials/intermediate/model_parallel_tutorial.html.
+
+    Usage:
+    >>> my_model = PipelineHelper().make_parallel(my_model)
+
+    Note that you will need to manually implement logic which handles the
+    moved layers.
+    """
+
+    def __init__(self):
+        self.__device_allocations = {}
+        self.num_devices = torch.cuda.device_count()
+        self.devices = []
+        for i in range(self.num_devices):
+            d = f'cuda:{i}'
+            self.devices.append(d)
+            self.__device_allocations[d] = 0
+
+    def make_parallel(self, model: torch.nn.Module) -> torch.nn.Module:
+        """
+        Allocate specific layers in a model to be ModelParallel.
+
+        Limited to only ModuleLists within the model.  Uses some heuristics to
+        attempt to evenly distribute layers across GPUs, in order to balance
+        memory usage. They are:
+
+        - Assume the 0th GPU will host the optimizer, word embeddings, etc.
+        - Assume activation memory is linear with the number of parameters.
+        - All layers are approximately equal in size.
+        """
+
+        # first assume all layers will go onto gpu 0 as an optimizer. The
+        # optimizer and embeddings are not quite as expensive as the
+        # activations (which scale via batchsize), Empirically, I found this
+        # heuristic works well enough.
+        self.__device_allocations['cuda:0'] += trainable_parameters(model) * 3
+
+        model.apply(self._place_modulelist)
+        return model
+
+    def _place_modulelist(self, submodule: torch.nn.Module) -> None:
+        if not isinstance(submodule, torch.nn.ModuleList):
+            # not a ModuleList, leave it untouched
+            return
+
+        assert isinstance(submodule, torch.nn.ModuleList)  # for typechecker
+        layers = submodule
+
+        # mark this section as MP
+        layers.is_model_parallel = True  # type: ignore
+
+        # next, let's figure out how many parameters we can assign to each GPU,
+        # but not make actual assignments yet. Assignments come later because we
+        # want consecutive layers to be collocated
```

Comment:
can you explain a bit more why we don't do the assignment at the same time?

Reply:
Essentially first pass tries to pack all the groups, and second pass assigns the groups to the layers. I couldn’t figure out how to do it in one pass.